File: backend/app/schemas/paper.py
"""Pydantic schemas for paper models."""
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class AuthorBase(BaseModel):
    """Base author schema."""
    full_name: str
    full_name_ru: Optional[str] = None
    affiliations: Optional[List[str]] = None
    affiliations_ru: Optional[List[str]] = None
    orcid: Optional[str] = None


class AuthorCreate(AuthorBase):
    """Schema for creating an author."""
    pass


class AuthorResponse(AuthorBase):
    """Schema for author response."""
    id: str
    
    class Config:
        from_attributes = True


class PaperBase(BaseModel):
    """Base paper schema."""
    title: str
    title_ru: Optional[str] = None
    abstract: Optional[str] = None
    abstract_ru: Optional[str] = None
    doi: Optional[str] = None
    arxiv_id: Optional[str] = None
    journal: Optional[str] = None
    journal_ru: Optional[str] = None
    publisher: Optional[str] = None
    volume: Optional[str] = None
    issue: Optional[str] = None
    pages: Optional[str] = None
    publication_date: Optional[datetime] = None
    publication_year: Optional[int] = None
    keywords: Optional[List[str]] = None
    keywords_ru: Optional[List[str]] = None
    language: str = "ru"


class PaperCreate(PaperBase):
    """Schema for creating a paper."""
    source_type: str = Field(..., description="Source system (eLibrary, CyberLeninka, etc.)")
    source_url: Optional[str] = None
    source_id: Optional[str] = None
    authors: List[AuthorCreate] = Field(default_factory=list)
    pdf_url: Optional[str] = None


class PaperResponse(PaperBase):
    """Schema for paper response."""
    id: str
    source_type: str
    source_url: Optional[str] = None
    citation_count: int = 0
    citation_count_rsci: int = 0
    authors: List[AuthorResponse] = Field(default_factory=list)
    crawled_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class PaperSearchRequest(BaseModel):
    """Schema for paper search request."""
    query: str = Field(..., min_length=1, description="Search query")
    limit: int = Field(20, ge=1, le=100)
    offset: int = Field(0, ge=0)
    filters: Optional[dict] = None


class SimilarPaperResponse(BaseModel):
    """Schema for similar paper response."""
    papers: List[PaperResponse]
    total: int


class IndexStats(BaseModel):
    """Schema for index statistics."""
    total_papers: int
    by_source: dict
    by_year: dict
    with_full_text: int
    processing_coverage: float

File: frontend/src/types/paper.ts
export interface Author {
  id: string
  full_name: string
  full_name_ru?: string
  affiliations?: string[]
  orcid?: string
}

export interface Paper {
  id: string
  title: string
  title_ru?: string
  abstract?: string
  abstract_ru?: string
  doi?: string
  arxiv_id?: string
  source_type: string
  source_url?: string
  journal?: string
  journal_ru?: string
  publisher?: string
  volume?: string
  issue?: string
  pages?: string
  publication_year?: number
  publication_date?: string
  keywords?: string[]
  keywords_ru?: string[]
  authors: Author[]
  citation_count: number
  citation_count_rsci: number
  pdf_url?: string
  language: string
  crawled_at: string
  updated_at: string
}

export interface SearchResponse {
  papers: Paper[]
  total: number
  limit: number
  offset: number
  search_type: string
}

export interface SearchFilters {
  year_from?: number
  year_to?: number
  source?: string
  journal?: string
  has_full_text?: boolean
  language?: string
}

export interface IndexStats {
  total_papers: number
  by_source: Record<string, number>
  by_year: Record<string, number>
  with_full_text: number
  processing_coverage: number
}
